src/connection.rs

```diff
@@ -0,0 +1,205 @@
+        let mut ctx = SslContext::builder(SslMethod::tls()).unwrap();
+        ctx.set_verify(SslVerifyMode::NONE);
+        ctx.set_certificate_chain_file(cert).unwrap();
+        ctx.set_private_key_file(&key, SslFiletype::PEM).unwrap();
+        ctx.set_cipher_list("ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES256-SHA:ECDHE-RSA-AES256-SHA:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA256:AES256-SHA256:AES128-SHA:AES256-SHA:ECDHE-ECDSA-DES-CBC3-SHA:ECDHE-RSA-DES-CBC3-SHA:DES-CBC3-SHA:!aNULL:!eNULL:!EXPORT:!DES:!RC4:!MD5:!PSK:!aECDH:!EDH-DSS-DES-CBC3-SHA:!EDH-RSA-DES-CBC3-SHA:!KRB5-DES-CBC3-SHA").unwrap();
+        Arc::new(ctx.build())
+    }
+
+    pub fn read(&mut self) -> Result<MessageType, std::io::Error> {
+        // The first two bytes of a message is the id/type of the message
+        let mut id_bytes = [0; 2];
+        self.stream
+            .ssl_read(&mut id_bytes)
+            .map_err(|e| e.into_io_error().unwrap())?;
+
+        let id = Cursor::new(id_bytes).read_u16::<BigEndian>().unwrap();
+
+        // The next four bytes specify the length of the message
+        let mut length_bytes = [0; 4];
+        self.stream
+            .ssl_read(&mut length_bytes)
+            .map_err(|e| e.into_io_error().unwrap())?;
+
+        let length = Cursor::new(length_bytes).read_u32::<BigEndian>().unwrap();
+
+        // The rest of the message is the payload (length bytes)
+        let mut payload = vec![0; length as usize];
+        self.stream.read_exact(&mut payload)?;
+
+        // Evaluate payload
+        let message = MessageType::from_raw(id, &payload);
+
+        // Handle special messages
+        match &message {
+            // Save the current session
+            ServerSync(data) => self.session = Some(data.get_session()),
+
+            // Keep a list of all users
+            UserState(data) => {
+                // Replace or append user
+                if let Some(index) = self
+                    .users
+                    .iter()
+                    .position(|u| u.get_session() == data.get_session())
+                {
+                    // If there is a new channel id we update it
+                    if data.has_channel_id() {
+                        self.users[index].set_channel_id(data.get_channel_id());
+                    }
+                } else {
+                    // Add user
+                    self.users.push(*data.clone());
+                }
+            }
+            // Ignore other messages
+            _ => {}
+        };
+
+        Ok(message)
+    }
+
+    pub fn send<M: Message + HasType>(&mut self, message: &M) {
+        // Create an stream to ouput this message
```

Comment:
I don't require it, but I would appreciate it if you could correct obvious spelling mistakes like these.